### PR TITLE
ADC Battery monitor support, but cleaner

### LINKED ISSUE #9: ADC for battery information

### PR DESCRIPTION
**Pull request by BillThePlatypus**
This adds support for both the ADC, as well as analog battery monitors. Unlike the last PR for this, it doesn't include a ton of unrelated commits. Oops.

Closes #9

### CODE REVIEW

File: examples/analog/main.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2017, James Jackson
```
Comment (superjax) [inline examples/analog/main.cpp]: You should take credit for this file (also change the year to 2020)

File: examples/battery_monitor/main.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017, James Jackson
```
Comment (superjax) [inline examples/battery_monitor/main.cpp]: update

File: include/analog_digital_converter.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
+#ifndef ADC_H
```
Comment (superjax) [inline include/analog_digital_converter.h:33]: needs copyright

File: include/analog_digital_converter.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  // The parameter, rank, is the return value from add_channel when the channel was added
+  /**
+   * @brief Reads a single channel.
+   * @details The reading is between 0 and RAW_READING_MAX. 0 represents 0V, and
+   * RAW_READING_MAX represents REFERENCE_VOLTAGE (typically 3.3V).
+   * Takes the rank of the channel as a parameter, which is returned from add_channel
+   * @param rank The rank of the channel to be read.
+   * @return
+   */
+  uint16_t read(uint8_t rank) const;
```
Comment (superjax) [inline include/analog_digital_converter.h]: is `rank` a term from the datasheet?  I feel like `index` or `channel_id` might be easier to understand

File: include/analog_digital_converter.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  static constexpr uint8_t CHANNEL_COUNT{16};
+
+private:
+  const ADCHardwareStruct *adc_def_;
+  bool is_initialized_{false};
+  uint8_t current_channels;
+
+  // These two constants are used for adjusting the ADC configuration when a channel is added
+  static constexpr uint32_t SQR1_L_MASK{~0xFF0FFFFF};
+  static constexpr uint8_t SQR1_L_OFFSET{20};
```
Comment (superjax) [inline include/analog_digital_converter.h]: Is there a more self-documenting name for these?

File: include/analog_pin.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
+#ifndef ANALOG_H
```
Comment (superjax) [inline include/analog_pin.h:32]: Needs copyright

File: include/analog_pin.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+   * @brief Get the most recent reading of the pin, directly from the ADC.
+   * See AnalogDigitalConverter.read for the details on interpreting this value.
+   * @return The most recent raw reading
+   */
+  uint16_t read_raw() const;
+
+private:
+  GPIO gpio_;
+  AnalogDigitalConverter *adc_;
+  uint8_t rank;
```
Comment (superjax) [inline include/analog_pin.h]: `rank_`

Could we use `channel_id_` here?

File: include/battery_monitor.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#ifndef BATTERYMONITOR_H
```
Comment (superjax) [inline include/battery_monitor.h:33]: copyright

File: include/battery_monitor.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+   */
+  float read_current() const;
+  /**
+   * @brief Sets the voltage multiplier for the battery monitor.
+   * @details Because the battery voltage is usually much too high for the ADC to read,
+   * analog battery monitors use a voltage divider or similar to reduce the voltage to a
+   * reasonable level. The voltage read by the ADC is multiplied by the voltage multiplier
+   * to get the actual reading. This number can be obtained from the spec sheet for the
+   * battery monitor, or through simple math for DIY monitors. This value is in volts per volt
+   * (i.e. unitless).
```
Comment (superjax) [inline include/battery_monitor.h:77]: could we have a little example formula here?

i.e

output = multiplier * observed voltage

File: include/battery_monitor.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+   * @param multiplier The new voltage multiplier
+   */
+  void set_voltage_multiplier(double multiplier);
+  /**
+   * @brief Sets the current multiplier for the battery monitor.
+   * @details Analog battery monitors which support current measurement generate a voltage
+   * proportional to the current through them. This voltage is read by the ADC, and then
+   * multiplied by the current multiplier to measure the current. This number can be obtained
+   * from the spec sheet for the battery monitor. This value is in amps per volt.
+   * @param multiplier
```
Comment (superjax) [inline include/battery_monitor.h:89]: same, write the formula here

File: include/system.h [hunk 11]
```diff
@@ -117,6 +118,24 @@ typedef struct {
 	uint32_t DMA_Tx_IT_Bit;
 } uart_hardware_struct_t;
 
+typedef struct
```
Comment (superjax) [inline include/system.h]: C++ style struct?

File: src/analog_digital_converter.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
+#include "analog_digital_converter.h"
```
Comment (superjax) [inline src/analog_digital_converter.cpp:33]: copyright

**Review by superjax — APPROVED** [review]
This is awesome!!  

Thanks so much.  Basically very minor details, and missing copyright.

I'm going to approve it, just make those changes.

File: include/analog_pin.h (hunk 7 above)
Comment (BillThePlatypus) [reply include/analog_pin.h]: Rank is the term used in the datasheet.

File: include/analog_digital_converter.h (hunk 4 above)
Comment (BillThePlatypus) [reply include/analog_digital_converter.h]: I renamed these to index.

File: include/analog_digital_converter.h (hunk 5 above)
Comment (BillThePlatypus) [reply include/analog_digital_converter.h]: Renamed to `SQR1_CHANNEL_COUNT_MASK` and `SQR1_CHANNEL_COUNT_OFFSET`.

**Pull request merged**